File: internal/participant/participant.go
package participant

import (
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/Georgecane/openhost/internal/capability"
	"github.com/Georgecane/openhost/internal/identity"
	"github.com/Georgecane/openhost/internal/resource"
)

var (
	ErrInvalidParticipant = errors.New("invalid participant")
	ErrInvalidTransition  = errors.New("invalid participant state transition")
	ErrNotActive           = errors.New("participant is not active")
)

type State string

const (
	StateJoining State = "joining"
	StateActive  State = "active"
	StateDraining State = "draining"
	StateLeft    State = "left"
)

type Participant struct {
	mu         sync.RWMutex
	ID         identity.Identity
	Capability capability.Capability
	Resources  resource.ResourceFragment
	State      State
	JoinedAt   time.Time
	UpdatedAt  time.Time
}

func New(id identity.Identity, cap capability.Capability, resources resource.ResourceFragment, now time.Time) (*Participant, error) {
	if err := id.Validate(); err != nil || id.Kind != identity.ParticipantKind {
		return nil, fmt.Errorf("%w: participant identity must be valid", ErrInvalidParticipant)
	}
	if err := cap.Validate(); err != nil {
		return nil, err
	}
	if err := resources.Validate(); err != nil {
		return nil, err
	}
	if resources.Empty() {
		return nil, fmt.Errorf("%w: resources must not be empty", ErrInvalidParticipant)
	}
	if now.IsZero() {
		return nil, fmt.Errorf("%w: timestamp must not be zero", ErrInvalidParticipant)
	}

	return &Participant{
		ID:         id,
		Capability: cap,
		Resources:  resources,
		State:      StateJoining,
		JoinedAt:   now,
		UpdatedAt:  now,
	}, nil
}

func (p *Participant) Activate(now time.Time) error {
	return p.transition(StateActive, now)
}

func (p *Participant) BeginDrain(now time.Time) error {
	return p.transition(StateDraining, now)
}

func (p *Participant) Leave(now time.Time) error {
	return p.transition(StateLeft, now)
}

type Snapshot struct {
	ID         identity.Identity
	Capability capability.Capability
	Resources  resource.ResourceFragment
	State      State
	JoinedAt   time.Time
	UpdatedAt  time.Time
}

func (p *Participant) Snapshot() Snapshot {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return Snapshot{
		ID:         p.ID,
		Capability: p.Capability,
		Resources:  p.Resources,
		State:      p.State,
		JoinedAt:   p.JoinedAt,
		UpdatedAt:  p.UpdatedAt,
	}
}

func (p *Participant) Offer() (resource.ResourceFragment, error) {
	p.mu.RLock()
	defer p.mu.RUnlock()

	if p.State != StateActive {
		return resource.ResourceFragment{}, ErrNotActive
	}
	return p.Resources, nil
}

func (p *Participant) transition(next State, now time.Time) error {
	if p == nil {
		return ErrInvalidParticipant
	}
	if now.IsZero() {
		return fmt.Errorf("%w: timestamp must not be zero", ErrInvalidParticipant)
	}

	p.mu.Lock()
	defer p.mu.Unlock()

	if !validTransition(p.State, next) {
		return fmt.Errorf("%w: %s -> %s", ErrInvalidTransition, p.State, next)
	}

	p.State = next
	p.UpdatedAt = now
	return nil
}

func validTransition(from, to State) bool {
	switch from {
	case StateJoining:
		return to == StateActive || to == StateLeft
	case StateActive:
		return to == StateDraining || to == StateLeft
	case StateDraining:
		return to == StateLeft
	case StateLeft:
		return false
	default:
		return false
	}
}

File: src/participant.rs
use crate::capability::Capability;
use crate::identity::{Identity, Kind};
use crate::resource::ResourceFragment;
use std::sync::RwLock;
use std::time::SystemTime;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum State { Joining, Active, Draining, Left }

#[derive(Debug, Clone)]
pub struct ParticipantSnapshot {
    pub id: Identity,
    pub capability: Capability,
    pub resources: ResourceFragment,
    pub state: State,
    pub joined_at: SystemTime,
    pub updated_at: SystemTime,
}

pub struct Participant {
    id: Identity,
    capability: Capability,
    resources: ResourceFragment,
    state: RwLock<State>,
    joined_at: SystemTime,
    updated_at: RwLock<SystemTime>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ParticipantError { Invalid, InvalidTransition, NotActive }

impl std::fmt::Display for ParticipantError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result { write!(f, "{self:?}") }
}
impl std::error::Error for ParticipantError {}

impl Participant {
    pub fn new(id: Identity, capability: Capability, resources: ResourceFragment, now: SystemTime) -> Result<Self, ParticipantError> {
        if id.kind != Kind::Participant || id.validate().is_err() || capability.validate().is_err()
            || resources.validate().is_err() || resources.is_empty() || now == SystemTime::UNIX_EPOCH
        { return Err(ParticipantError::Invalid); }
        Ok(Self { id, capability, resources, state: RwLock::new(State::Joining), joined_at: now, updated_at: RwLock::new(now) })
    }

    pub fn activate(&self, now: SystemTime) -> Result<(), ParticipantError> { self.transition(State::Active, now) }
    pub fn begin_drain(&self, now: SystemTime) -> Result<(), ParticipantError> { self.transition(State::Draining, now) }
    pub fn leave(&self, now: SystemTime) -> Result<(), ParticipantError> { self.transition(State::Left, now) }

    fn transition(&self, next: State, now: SystemTime) -> Result<(), ParticipantError> {
        if now == SystemTime::UNIX_EPOCH { return Err(ParticipantError::Invalid); }
        let mut state = self.state.write().expect("participant state lock poisoned");
        let valid = matches!((*state, next),
            (State::Joining, State::Active) | (State::Joining, State::Left)
            | (State::Active, State::Draining) | (State::Active, State::Left)
            | (State::Draining, State::Left));
        if !valid { return Err(ParticipantError::InvalidTransition); }
        *state = next;
        *self.updated_at.write().expect("participant timestamp lock poisoned") = now;
        Ok(())
    }

    pub fn snapshot(&self) -> ParticipantSnapshot {
        ParticipantSnapshot {
            id: self.id.clone(),
            capability: self.capability,
            resources: self.resources,
            state: *self.state.read().expect("participant state lock poisoned"),
            joined_at: self.joined_at,
            updated_at: *self.updated_at.read().expect("participant timestamp lock poisoned"),
        }
    }

    pub fn offer(&self) -> Result<ResourceFragment, ParticipantError> {
        if *self.state.read().expect("participant state lock poisoned") != State::Active {
            return Err(ParticipantError::NotActive);
        }
        Ok(self.resources)
    }
}
